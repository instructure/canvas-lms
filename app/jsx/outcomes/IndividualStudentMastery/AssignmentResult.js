/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import React from 'react'
import I18n from 'i18n!outcomes'
import _ from 'lodash'
import Flex, { FlexItem } from '@instructure/ui-layout/lib/components/Flex'
import View from '@instructure/ui-layout/lib/components/View'
import Link from '@instructure/ui-elements/lib/components/Link'
import Text from '@instructure/ui-elements/lib/components/Text'
import IconAssignment from '@instructure/ui-icons/lib/Line/IconAssignment'
import IconQuiz from '@instructure/ui-icons/lib/Line/IconQuiz'
import * as shapes from './shapes'
import Ratings from '../../rubrics/Ratings'

const AssignmentResult = ({ outcome, result, outcomeProficiency }) => {
  const { ratings } = outcome
  const { html_url: url, name, submission_types: types } = result.assignment
  const isQuiz = types && types.indexOf('online_quiz') >= 0

  return (
    <Flex padding="small" direction="column" alignItems="stretch">
      <FlexItem>
        <Link href={ url }>
          <Flex alignItems="center">
            <FlexItem><Text size="medium">
              {
                isQuiz ? <IconQuiz /> : <IconAssignment />
              }
            </Text></FlexItem>
            <FlexItem padding="0 x-small"><Text weight="bold">{ name }</Text></FlexItem>
          </Flex>
        </Link>
      </FlexItem>
      <FlexItem padding="x-small 0">
        <View padding="x-small 0 0 0">
          <Text size="small" fontStyle="italic" weight="bold">{
            result.hide_points ? I18n.t('Your score') : I18n.t('Your score: %{score}', { score: result.score })
          }</Text>
        </View>
      </FlexItem>
      <FlexItem borderWidth="small">
        <div className="react-rubric">
          <div className="ratings">
            <Ratings
              tiers={ratings}
              points={result.score}
              hidePoints={result.hide_points}
              useRange={false}
              customRatings={_.get(outcomeProficiency, 'ratings')}
              defaultMasteryThreshold={outcome.mastery_points}
              pointsPossible={outcome.points_possible}
              assessing={false}
              isSummary={false}
            />
          </div>
        </div>
      </FlexItem>
    </Flex>
  )
}

AssignmentResult.propTypes = {
  result: shapes.outcomeResultShape.isRequired,
  outcome: shapes.outcomeShape.isRequired,
  outcomeProficiency: shapes.outcomeProficiencyShape
}

AssignmentResult.defaultProps = {
  outcomeProficiency: null
}

export default AssignmentResult
