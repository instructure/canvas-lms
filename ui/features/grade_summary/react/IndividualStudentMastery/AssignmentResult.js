/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import React from 'react'
import I18n from 'i18n!IndividiualStudentMasteryAssignmentResult'
import _ from 'lodash'
import {View} from '@instructure/ui-view'
import {Flex} from '@instructure/ui-flex'
import {Button} from '@instructure/ui-buttons'
import {Text} from '@instructure/ui-text'
import {IconAssignmentLine, IconQuizLine, IconHighlighterLine} from '@instructure/ui-icons'
import * as shapes from './shapes'
import Ratings from '@canvas/rubrics/react/Ratings'

const scoreFromPercent = (percent, outcome) => {
  if (outcome.points_possible > 0) {
    return +(percent * outcome.points_possible).toFixed(2)
  } else {
    return +(percent * outcome.mastery_points).toFixed(2)
  }
}

const scaleScore = (score, possible, outcome) => {
  if (!possible) return score
  if (outcome.points_possible > 0) {
    return +((score / possible) * outcome.points_possible).toFixed(2)
  } else {
    return +((score / possible) * outcome.mastery_points).toFixed(2)
  }
}

const renderLinkedResult = (name, url, isQuiz) => (
  <Button
    variant="link"
    href={url}
    theme={{mediumPaddingHorizontal: '0', mediumHeight: 'normal', fontWeight: '700'}}
    icon={isQuiz ? IconQuizLine : IconAssignmentLine}
  >
    {name}
  </Button>
)

const renderUnlinkedResult = name => (
  <Flex alignItems="center">
    <Flex.Item>
      <Text size="medium">
        <IconHighlighterLine />
      </Text>
    </Flex.Item>
    <Flex.Item padding="0 x-small">
      <Text weight="bold">{name}</Text>
    </Flex.Item>
  </Flex>
)

const AssignmentResult = ({outcome, result, outcomeProficiency}) => {
  const {ratings} = outcome
  const {html_url: url, name, submission_types: types} = result.assignment
  const isQuiz = types && types.indexOf('online_quiz') >= 0
  const score = result.percent
    ? scoreFromPercent(result.percent, outcome)
    : scaleScore(result.score, result.points_possible, outcome)
  return (
    <Flex padding="small" direction="column" alignItems="stretch">
      <Flex.Item>
        {url.length > 0 ? renderLinkedResult(name, url, isQuiz) : renderUnlinkedResult(name)}
      </Flex.Item>
      <Flex.Item padding="x-small 0">
        <View padding="x-small 0 0 0">
          <Text size="small" fontStyle="italic" weight="bold">
            {result.hide_points ? I18n.t('Your score') : I18n.t('Your score: %{score}', {score})}
          </Text>
        </View>
      </Flex.Item>
      <Flex.Item borderWidth="small">
        <div className="react-rubric">
          <div className="ratings">
            <Ratings
              tiers={ratings}
              points={score}
              hidePoints={result.hide_points}
              useRange={false}
              customRatings={_.get(outcomeProficiency, 'ratings')}
              defaultMasteryThreshold={outcome.mastery_points}
              pointsPossible={outcome.points_possible}
              assessing={false}
              isSummary={false}
            />
          </div>
        </div>
      </Flex.Item>
    </Flex>
  )
}

AssignmentResult.propTypes = {
  result: shapes.outcomeResultShape.isRequired,
  outcome: shapes.outcomeShape.isRequired,
  outcomeProficiency: shapes.outcomeProficiencyShape
}

AssignmentResult.defaultProps = {
  outcomeProficiency: null
}

export default AssignmentResult
