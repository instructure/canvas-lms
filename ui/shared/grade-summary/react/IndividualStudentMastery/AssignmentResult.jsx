/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import React from 'react'
import {useScope as createI18nScope} from '@canvas/i18n'
import _ from 'lodash'
import {View} from '@instructure/ui-view'
import {Flex} from '@instructure/ui-flex'
import {Text} from '@instructure/ui-text'
import {IconAssignmentLine, IconQuizLine, IconHighlighterLine} from '@instructure/ui-icons'
import * as shapes from './shapes'
import Ratings from '@canvas/rubrics/react/Ratings'
import {scoreFromPercent, scaleScore} from './scoreCalculation'

import {Link} from '@instructure/ui-link'

const I18n = createI18nScope('IndividiualStudentMasteryAssignmentResult')

const renderLinkedResult = (name, url, isQuiz) => (
  <Link
    href={url}
    isWithinText={false}
    themeOverride={{mediumPaddingHorizontal: '0', mediumHeight: 'normal', fontWeight: '700'}}
    renderIcon={isQuiz ? IconQuizLine : IconAssignmentLine}
  >
    {name}
  </Link>
)

const renderUnlinkedResult = name => (
  <Flex alignItems="center">
    <Flex.Item>
      <Text size="medium">
        <IconHighlighterLine />
      </Text>
    </Flex.Item>
    <Flex.Item padding="0 x-small">
      <Text weight="bold">{name}</Text>
    </Flex.Item>
  </Flex>
)

const AssignmentResult = ({outcome, result, outcomeProficiency}) => {
  const ratings = result.outcomeRatingsFromRubric || outcome.ratings
  const pointsPossibleFromOutcomeRatingsFromRubric = result.outcomeRatingsFromRubric
  ? Math.max(...ratings.map(rating => rating.points)) || null
  : null
  const {html_url: url, name, submission_types: types} = result.assignment
  const isQuiz = types && types.indexOf('online_quiz') >= 0
  const score = result.percent
    ? scoreFromPercent(result.percent, outcome, pointsPossibleFromOutcomeRatingsFromRubric)
    : scaleScore(result.score, result.points_possible, outcome, pointsPossibleFromOutcomeRatingsFromRubric)
  return (
    <Flex padding="small" direction="column" alignItems="stretch">
      <Flex.Item>
        {url.length > 0 ? renderLinkedResult(name, url, isQuiz) : renderUnlinkedResult(name)}
      </Flex.Item>
      <Flex.Item padding="x-small 0">
        <View padding="x-small 0 0 0">
          <Text size="small" fontStyle="italic" weight="bold">
            {result.hide_points ? I18n.t('Your score') : I18n.t('Your score: %{score}', {score})}
          </Text>
        </View>
      </Flex.Item>
      <Flex.Item borderWidth="small">
        <div className="react-rubric">
          <div className="ratings">
            <Ratings
              tiers={ratings}
              points={score}
              hidePoints={result.hide_points}
              useRange={false}
              customRatings={_.get(outcomeProficiency, 'ratings')}
              defaultMasteryThreshold={outcome.mastery_points}
              pointsPossible={outcome.points_possible}
              assessing={false}
              isSummary={false}
            />
          </div>
        </div>
      </Flex.Item>
    </Flex>
  )
}

AssignmentResult.propTypes = {
  result: shapes.outcomeResultShape.isRequired,
  outcome: shapes.outcomeShape.isRequired,
  outcomeProficiency: shapes.outcomeProficiencyShape,
}

AssignmentResult.defaultProps = {
  outcomeProficiency: null,
}

export default AssignmentResult
