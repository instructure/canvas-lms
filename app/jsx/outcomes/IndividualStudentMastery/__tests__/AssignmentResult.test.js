/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import React from 'react'
import { render, shallow } from 'enzyme'
import AssignmentResult from '../AssignmentResult'

const defaultProps = (props = {}) => (
  Object.assign({
    outcome: {
      id: 1,
      mastered: false,
      ratings: [
        { description: 'My first rating' },
        { description: 'My second rating' }
      ],
      results: [],
      title: 'foo',
      mastery_points: 3,
      points_possible: 5
    },
    result: {
      id: 1,
      score: 1,
      percent: 0.1,
      assignment: {
        id: 1,
        html_url: 'http://foo',
        name: 'My assignment',
        submission_types: 'online_quiz'
      }
    }
  }, props)
)

it('renders the AlignmentResult component', () => {
  const wrapper = shallow(<AssignmentResult {...defaultProps()}/>)
  expect(wrapper.debug()).toMatchSnapshot()
})

it('includes the assignment name', () => {
  const wrapper = render(<AssignmentResult {...defaultProps()}/>)
  expect(wrapper.text()).toMatch('My assignment')
})

it('includes the ratings of the outcome', () => {
  const wrapper = render(<AssignmentResult {...defaultProps()}/>)
  expect(wrapper.text()).toMatch('My second rating')
})

it('shows scores when points are not hidden', () => {
  const wrapper = render(<AssignmentResult {...defaultProps()}/>)
  expect(wrapper.text()).toMatch('Your score: 1')
})

it('does not show scores when points are hidden', () => {
  const props = defaultProps()
  props.result.hide_points = true
  const wrapper = render(<AssignmentResult {...props}/>)
  expect(wrapper.text()).toMatch('Your score')
  expect(wrapper.text()).not.toMatch('Your score: 1')
})
