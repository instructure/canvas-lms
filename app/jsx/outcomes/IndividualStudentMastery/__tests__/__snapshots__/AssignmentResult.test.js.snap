// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the AlignmentResult component 1`] = `
"<Flex padding=\\"small\\" direction=\\"column\\" alignItems=\\"stretch\\" as=\\"span\\" justifyItems=\\"start\\" inline={false} visualDebug={false} wrapItems={false}>
  <FlexItem as=\\"span\\" grow={false} shrink={false}>
    <Link href=\\"http://foo\\" variant=\\"default\\" as=\\"button\\" linkRef={[Function]} ellipsis={false} iconPlacement=\\"start\\">
      <Flex alignItems=\\"center\\" as=\\"span\\" direction=\\"row\\" justifyItems=\\"start\\" inline={false} visualDebug={false} wrapItems={false}>
        <FlexItem as=\\"span\\" grow={false} shrink={false}>
          <Text size=\\"medium\\" as=\\"span\\" letterSpacing=\\"normal\\">
            <IconQuiz />
          </Text>
        </FlexItem>
        <FlexItem padding=\\"0 x-small\\" as=\\"span\\" grow={false} shrink={false}>
          <Text weight=\\"bold\\" as=\\"span\\" size=\\"medium\\" letterSpacing=\\"normal\\">
            My assignment
          </Text>
        </FlexItem>
      </Flex>
    </Link>
  </FlexItem>
  <FlexItem padding=\\"x-small 0\\" as=\\"span\\" grow={false} shrink={false}>
    <View padding=\\"x-small 0 0 0\\" display=\\"auto\\">
      <Text size=\\"small\\" fontStyle=\\"italic\\" weight=\\"bold\\" as=\\"span\\" letterSpacing=\\"normal\\">
        Your score: 1
      </Text>
    </View>
  </FlexItem>
  <FlexItem borderWidth=\\"small\\" as=\\"span\\" grow={false} shrink={false}>
    <div className=\\"react-rubric\\">
      <div className=\\"ratings\\">
        <Ratings tiers={{...}} points={1} hidePoints={false} useRange={false} customRatings={[undefined]} defaultMasteryThreshold={3} pointsPossible={5} assessing={false} isSummary={false} footer={{...}} onPointChange={[Function]} />
      </div>
    </div>
  </FlexItem>
</Flex>"
`;
